### ISSUE TITLE
ServiceCollection.Configure<ConcurrentDictionary<string, T>>(IConfigurationSection) fails in NET7

### ISSUE THREAD

**Issue by matthewDDennis** [dotnet/runtime#78994]
### Description

I have an **ModuleCollection** class which is 
``` c#
   /// <summary>
    /// The set of modules for backend processing.
    /// </summary>
    public class ModuleCollection : ConcurrentDictionary<string, ModuleConfig>
    {
        /// <summary>
        /// This constructor allows our modules collection to be case insensitive on the key.
        /// </summary>
        public ModuleCollection() : base(StringComparer.OrdinalIgnoreCase) { }
    }
```
In NET6
```c#
            services.Configure<ModuleCollection>(configuration.GetSection("Modules"));
```
would load the **ModuleCollection** with the **ModuleConfig**s from multiple `modulesettings.json` files that had been registered with the `ConfigurationBuilder`. The` IOption<ModuleCollection> `injected would have the populated collection as its value.

In NET7 the `IOption<ModuleCollection>.Value` is an empty `ModuleCollection`

I worked around this by
```c#
            services.AddOptions<ModuleCollection>()
                    .Configure(moduleCollection =>
                    {
                        var moduleNames = configuration.GetSection("Modules").GetChildren().Select(x => x.Key).ToList();
                        foreach (var moduleName in moduleNames)
                        {
                            if (moduleName is not null)
                            {
                                ModuleConfig moduleConfig = new ModuleConfig();
                                configuration.Bind($"Modules:{moduleName}", moduleConfig);
                                moduleCollection.TryAdd(moduleName, moduleConfig);
                            }
                        }
                    });
```

### Reproduction Steps

see above

### Expected behavior

Same as NET6
`IOption<ModuleCollection>.Value` contains the collection of `ModuleConfig`s/

### Actual behavior

The `IOption<ModuleCollection>.Value` is an empty collection.

### Regression?

It worked on NET6

### Known Workarounds

```c#
            ModuleCollection moduleCollection = new();
            services.AddOptions<ModuleCollection>()
                    .Configure(moduleCollection =>
                    {
                        var moduleNames = configuration.GetSection("Modules").GetChildren().Select(x => x.Key).ToList();
                        foreach (var moduleName in moduleNames)
                        {
                            if (moduleName is not null)
                            {
                                ModuleConfig moduleConfig = new ModuleConfig();
                                configuration.Bind($"Modules:{moduleName}", moduleConfig);
                                moduleCollection.TryAdd(moduleName, moduleConfig);
                            }
                        }
                    });
```

### Configuration

NET7, VS2022 17.4.1

### Other information

I can bind to a single `ModuleConfig`, just the binding to the dictionary is failing, so it is probably in the Dictionary Binder in NET7.

**Comment by matthewDDennis** [comment]
Probably should be labeled **area-Extensions-Options**

**Comment by RichardD2** [comment]
Do you actually need the `ModuleCollection moduleCollection = new();` line? That variable doesn't seem to be used anywhere else - the other references use the lambda method parameter with the same name.

**Comment by matthewDDennis** [comment]
> Do you actually need the `ModuleCollection moduleCollection = new();` line? That variable doesn't seem to be used anywhere else - the other references use the lambda method parameter with the same name.

You are correct.  That was a copy & paste error from my test code to determine what was broken.
I've updated my question.

Thanks.

**Comment by tarekgh** [comment]
@matthewDDennis could you please try to run using `Microsoft.Extensions.*` packages from the nuget private feed 

```Xml
    <add key="dotnet8" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json" />
```

and use the version: `8.0.0-alpha.1.22578.1` 

I am asking because this look it is already fixed by https://github.com/dotnet/runtime/pull/78118. 

If it is still reproduced, please provide a small app that reproduces the problem we can try. Thanks.

**Comment by matthewDDennis** [comment]
> 

I am not able to do that at the moment as I am working on the migration from net6 to net7.
I still have some work to go before I will be able to branch and check that out.
I see you have back ported the fix to 7.0.1. 
Is that available in the standard NuGet feed yet?

**Comment by tarekgh** [comment]
Either you can wait for .NET 7.0.1 release, or you try what I mentioned in the comment https://github.com/dotnet/runtime/issues/78994#issuecomment-1332798171 or download the file https://dotnetbuilds.azureedge.net/public/aspnetcore/Runtime/7.0.1-servicing.22565.7/aspnetcore-runtime-7.0.1-win-x64.zip and copy the assembly `Microsoft.Extensions.Configuration.Binder.dll` to your `dotnet\shared\Microsoft.AspNetCore.App\7.0.0` folder. Make sure to back up the original file before overwriting it so you may copy it back.

**Comment by tarekgh** [comment]
I have confirmed this issue is fixed in `7.0.1`. I used the following code. I can see the issue repro on `7.0.0`.

### repro code

```C#
        public class ModuleCollection : ConcurrentDictionary<string, string>
        {
            public ModuleCollection() : base(StringComparer.OrdinalIgnoreCase) { }
        }


        var configuration = new ConfigurationBuilder()
            .AddJsonFile("appsettings.json")
            .Build();

        var services = new ServiceCollection();
        services.Configure<ModuleCollection>(configuration.GetSection("Modules"));

        var val = services.BuildServiceProvider().GetRequiredService<IOptions<ModuleCollection>>().Value;
```

### appsettings.json

```Json
{
    "Modules": {
        "module1": "moduleConfig1",
        "module2": "moduleConfig2",
        "module3": "moduleConfig3"
    }
}
```
